src/main/java/org/folio/circulation/services/EventPublishingService.java

```diff
@@ -0,0 +1,164 @@
+      write(payloadJsonObject, "returnDate", loan.getReturnDate());
+
+      publishEvent(ITEM_CHECKED_IN_EVENT_TYPE, payloadJsonObject.encode());
+    }
+    else {
+      logger.error("Failed to publish {} event: loan is null", ITEM_CHECKED_IN_EVENT_TYPE);
+    }
+
+    return completedFuture(succeeded(checkInProcessRecords));
+  }
+
+  public CompletableFuture<Result<Loan>> publishDeclaredLostEvent(Loan loan) {
+    if (loan != null) {
+      JsonObject payloadJsonObject = new JsonObject();
+      write(payloadJsonObject, "userId", loan.getUserId());
+      write(payloadJsonObject, "loanId", loan.getId());
+
+      publishEvent(ITEM_DECLARED_LOST_EVENT_TYPE, payloadJsonObject.encode());
+    }
+
+    return completedFuture(succeeded(loan));
+  }
+
+  private CompletableFuture<Result<Loan>> publishDueDateChangedEvent(Loan loan) {
+    if (loan != null) {
+      JsonObject payloadJsonObject = new JsonObject();
+      write(payloadJsonObject, "userId", loan.getUserId());
+      write(payloadJsonObject, "loanId", loan.getId());
+      write(payloadJsonObject, "dueDate", loan.getDueDate());
+      write(payloadJsonObject, "dueDateChangedByRecall", loan.wasDueDateChangedByRecall());
+
+      publishEvent(LOAN_DUE_DATE_UPDATED_EVENT_TYPE, payloadJsonObject.encode());
+    }
+
+    return completedFuture(succeeded(loan));
+  }
+
+  public CompletableFuture<Result<LoanAndRelatedRecords>> publishDueDateChangedEvent(
+    LoanAndRelatedRecords loanAndRelatedRecords) {
+
+    if (loanAndRelatedRecords.getLoan() != null) {
+      Loan loan = loanAndRelatedRecords.getLoan();
+      publishDueDateChangedEvent(loan);
+    }
+
+    return completedFuture(succeeded(loanAndRelatedRecords));
+  }
+
+  public CompletableFuture<Result<RequestAndRelatedRecords>> publishDueDateChangedEvent(
+    RequestAndRelatedRecords requestAndRelatedRecords) {
+
+    LoanRepository loanRepository = new LoanRepository(clients);
+    loanRepository.findOpenLoanForRequest(requestAndRelatedRecords.getRequest())
+      .thenCompose(r -> r.after(this::publishDueDateChangedEvent));
+
+    return completedFuture(succeeded(requestAndRelatedRecords));
+  }
+
+  private void publishEvent(String eventType, String payload) {
+    Event event = new Event()
+      .withId(UUID.randomUUID().toString())
+      .withEventType(eventType)
+      .withEventPayload(payload)
+      .withEventMetadata(new EventMetadata()
+        .withPublishedBy(PubSubClientUtils.constructModuleName())
+        .withTenantId(okapiHeaders.get(OKAPI_TENANT_HEADER))
+        .withEventTTL(1));
+
+    OkapiConnectionParams params = new OkapiConnectionParams(okapiHeaders, vertxContext.owner());
+
+    PubSubClientUtils.sendEventMessage(event, params)
+      .whenComplete((result, throwable) -> {
+        if (result) {
+          logger.debug("Event published successfully: {}", event.getId());
+        } else {
+          logger.error("Failed to publish event: {}", event.getId());
```

Comment:
Should we log the `throwable` as well?

Reply:
Added throwable, event type and payload.